release/pkg/manager/calico/manager.go

```diff
@@ -552,13 +555,68 @@ type imageExistsResult struct {
 	err    error
 }
 
-func imgExists(name string, component registry.Component, ch chan imageExistsResult) {
-	r := imageExistsResult{
-		name:  name,
-		image: component.String(),
+// checkHashreleaseImagesPublished checks that the images required for the hashrelease exist in the specified registries.
+func (r *CalicoManager) checkHashreleaseImagesPublished() ([]registry.Component, error) {
+	logrus.Info("Checking images required for hashrelease have already been published")
+	timeout := 30 * time.Second
+
+	numOfComponents := len(r.imageComponents)
+	if numOfComponents == 0 {
+		logrus.Info("No images to check")
+		return nil, fmt.Errorf("no images to check")
 	}
-	r.exists, r.err = registry.ImageExists(component.ImageRef())
-	ch <- r
+
+	var wg sync.WaitGroup
+	resultsCh := make(chan imageExistsResult, numOfComponents)
+
+	for name, component := range r.imageComponents {
+		wg.Add(1)
+		go func(name string, component registry.Component, ch chan imageExistsResult) {
+			defer wg.Done()
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+
+			resultCh := make(chan imageExistsResult, 1)
+
+			go func() {
```

Comment:
Like we discussed in Slack, I'd like to see this simplified a lot. I think this is probably more concurrency / complexity than we need for this task, and will be prone to future bugs.

I think we can keep the previous implementation here and revert this change, and then make the change suggested here: https://github.com/projectcalico/calico/pull/9681/files#r1907632859